In [1]:
import numpy as np
import matplotlib.pyplot as plt
import tensorflow as tf
from tensorflow.keras import layers, models
from sklearn.metrics import confusion_matrix 

Loading the images and labels, then adjusting so that the data is in an easier format to work with

In [2]:
Train_images = np.load('trnImage.npy')
Train_labels = np.load('trnLabel.npy')
Test_images = np.load('tstImage.npy')
Test_labels = np.load('tstLabel.npy')
Train_images = np.moveaxis(Train_images, 3, 0) 
Test_images = np.moveaxis(Test_images, 3, 0) 
Train_labels = Train_labels.reshape(-1,)
Test_labels = Test_labels.reshape(-1,)

This method is for ploting images

In [3]:
def plotImage(images, labels, pos):
    plt.figure(figsize = (15,2))
    plt.imshow(images[:,:,:,pos])

This is the artificial neural network, the values for number of neuron were decided upon through trial and error

In [4]:
ann = models.Sequential([
        layers.Flatten(input_shape=(32,32,3)),
        layers.Dense(2000, activation='relu'),
        layers.Dense(500, activation='relu'),
        layers.Dense(10, activation='sigmoid')    
    ])

ann.compile(optimizer='SGD',
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

ann.fit(Train_images, Train_labels, epochs=5)

Epoch 1/5
313/313 [==============================] - 15s 49ms/step - loss: 2.0593 - accuracy: 0.24460s - loss: 2.0624 - 
Epoch 2/5
313/313 [==============================] - 15s 49ms/step - loss: 1.8697 - accuracy: 0.3338
Epoch 3/5
313/313 [==============================] - 15s 49ms/step - loss: 1.7854 - accuracy: 0.3601
Epoch 4/5
313/313 [==============================] - 15s 48ms/step - loss: 1.7317 - accuracy: 0.3853
Epoch 5/5
313/313 [==============================] - 15s 49ms/step - loss: 1.6804 - accuracy: 0.4079


Testing the ANN on the test set of images

In [5]:
pred = ann.evaluate(Test_images, Test_labels)

32/32 [==============================] - 1s 17ms/step - loss: 1.8221 - accuracy: 0.3480


CNN accuracy of 34.8%

The convolutional neural network with the first layer having 40 filters and the second having 80, using 
Max Pooling for dimensionality reduction

In [12]:
cnn = models.Sequential([
    layers.Conv2D(filters=40, kernel_size=(3, 3), activation='relu', input_shape=(32, 32, 3)),
    layers.MaxPooling2D((2, 2)),
    layers.Conv2D(filters=80, kernel_size=(3, 3), activation='relu'),
    layers.MaxPooling2D((2, 2)),
    layers.Flatten(),
    layers.Dense(60, activation='relu'),
    layers.Dense(10, activation='softmax')
])
cnn.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

cnn.fit(Train_images, Train_labels, epochs=5)

Epoch 1/5
313/313 [==============================] - 13s 43ms/step - loss: 1.8443 - accuracy: 0.3246
Epoch 2/5
313/313 [==============================] - 10s 31ms/step - loss: 1.4694 - accuracy: 0.4675
Epoch 3/5
313/313 [==============================] - 13s 43ms/step - loss: 1.3208 - accuracy: 0.5325
Epoch 4/5
313/313 [==============================] - 13s 43ms/step - loss: 1.1951 - accuracy: 0.5776
Epoch 5/5
313/313 [==============================] - 13s 43ms/step - loss: 1.1158 - accuracy: 0.6057


Evaluating the CNN on the test data

In [15]:
cnn.evaluate(Test_images, Test_labels)

32/32 [==============================] - 0s 13ms/step - loss: 1.2454 - accuracy: 0.5710


An accuracy of 57.1%

In [35]:
pred = cnn.predict(Test_images)
classes = [np.argmax(element) for element in pred]


Confusion matrix

In [34]:
confusion_matrix(classes, Test_labels)

array([[67, 10,  6,  2,  2,  0,  0,  1, 21,  6],
       [ 3, 68,  0,  0,  0,  1,  0,  0,  6, 21],
       [ 6,  0, 36,  8,  3,  3,  2,  1,  2,  0],
       [ 3,  2,  9, 39,  6, 19,  6, 11,  2,  4],
       [ 7,  4, 19, 17, 66, 12, 14, 15,  6,  1],
       [ 2,  3, 10, 16,  3, 54,  3, 10,  0,  5],
       [ 3,  3, 10, 12, 11,  5, 73,  1,  0,  6],
       [ 0,  2,  9,  4,  5,  5,  2, 58,  0,  3],
       [ 8,  5,  1,  0,  3,  0,  0,  1, 61,  5],
       [ 1,  3,  0,  2,  1,  1,  0,  2,  2, 49]], dtype=int64)